### PR TITLE
feat: prune idle volume browser helper containers

### LINKED ISSUE #2765: ⚡️ Feature: Volume helpers get orphaned
### Feature Description

The volume Browser tab (/volumes/[volumeName] → Browser) browses a volume's contents read-only. To do this the backend spins up a Docker helper container that mounts the volume read-only at /volume. These volume helpers should have a defined lifecycle so they are turned off when no longer in use.

### Problem It Solves

**The problem:** these helpers leak. If the user closes the tab, loses network, or the browser crashes, nothing tells the backend to tear the helper down — it keeps running and holds the volume, which can block volume deletes/prunes and waste resources.

### Proposed Solution

 layered cleanup: eager client-side teardown when the user leaves the browser, plus a backend reaper job as a guaranteed backstop that
  removes any helper left idle past a configurable timeout.

  How it works

  Helper lifecycle
  - Created per volume on the first read-only browse request; tracked in helperByVolume[volumeName] with a lastUsedAt timestamp.
  - Subsequent requests reuse the same helper and refresh lastUsedAt. 
  - Helpers are independent per volume — no "one at a time" global constraint.

  Teardown paths (layered, intentional)
  1. Eager client-side (VolumeBrowser.svelte):
    - onDestroy → stopBrowsing(volumeName) on tab switch, in-app nav, back button, and opening another volume (unmount).
    - pagehide → navigator.sendBeacon(.../browse/stop) on hard close/refresh.
  2. Manual — "Stop browsing" button → same stopBrowsing.
  3. Backend reaper (backstop) — VolumeHelperReaperJob runs every minute (0 * * * * *, fixed), removing any helper idle ≥ the timeout. Catches leaks
  the client paths miss.
  4. Implicit — DeleteVolume / PruneVolumes stop helpers first so Docker doesn't see the volume as in-use.

  ▎ Design note: teardown is eager, not lazy. The helper dies the moment you leave the browser; the 10-min idle timeout is the safety net, not the 
  ▎ primary mechanism.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you be willing to work on this feature?

- [x] Yes, I'd like to implement this feature
- [ ] I could help with parts of this feature
- [ ] No, I'm just suggesting the feature

### PR DESCRIPTION
**Pull request by Zgrill2**
## Checklist

- [x] This PR is **not** opened from my fork’s `main` branch

## What This PR Implements

Volume Browser Helper Reaper

  Context / Motivation

  The volume Browser tab (/volumes/[volumeName] → Browser) browses a volume's contents read-only. To do this the backend spins up a Docker helper
  container that mounts the volume read-only at /volume

  The problem: these helpers leak. If the user closes the tab, loses network, or the browser crashes, nothing tells the backend to tear the
  helper down — it keeps running and holds the volume, which can block volume deletes/prunes and waste resources.

  This PR adds layered cleanup: eager client-side teardown when the user leaves the browser, plus a backend reaper job as a guaranteed backstop that
  removes any helper left idle past a configurable timeout.

  

Fixes: [#2765](https://github.com/getarcaneapp/arcane/issues/2765)

## Changes Made

How it works

  Helper lifecycle
  - Created per volume on the first read-only browse request; tracked in helperByVolume[volumeName] with a lastUsedAt timestamp.
  - Subsequent requests reuse the same helper and refresh lastUsedAt. 
  - Helpers are independent per volume — no "one at a time" global constraint.

  Teardown paths (layered, intentional)
  1. Eager client-side (VolumeBrowser.svelte):
    - onDestroy → stopBrowsing(volumeName) on tab switch, in-app nav, back button, and opening another volume (unmount).
    - pagehide → navigator.sendBeacon(.../browse/stop) on hard close/refresh.
  2. Manual — "Stop browsing" button → same stopBrowsing.
  3. Backend reaper (backstop) — VolumeHelperReaperJob runs every minute (0 * * * * *, fixed), removing any helper idle ≥ the timeout. Catches leaks
  the client paths miss.
  4. Implicit — DeleteVolume / PruneVolumes stop helpers first so Docker doesn't see the volume as in-use.

  ▎ Design note: teardown is eager, not lazy. The helper dies the moment you leave the browser; the 10-min idle timeout is the safety net, not the primary mechanism.

## Testing Done

- [x] Development environment started: `./scripts/development/dev.sh start`
- [x] Frontend verified at http://localhost:3000
- [x] Backend verified at http://localhost:3552
- [x] Manual testing completed (describe):
- [x] No linting errors (e.g., `just lint all`)
- [x] Backend tests pass: `just test backend`

## AI Tool Used (if applicable)

AI Tool: Claude Code
Assistance Level: Moderate
What AI helped with: Design implementation plan, doing the actual code, running the linting and just tests.
I reviewed and edited all AI-generated output: Yes always. I had the design and all implementation plans were reviewed by me
I ran all required tests and manually verified changes: Yes, tested on my local laptop

## Additional Context
My partner uses Arcane and really wants this feature

**Disclaimer** Greptiles Reviews use AI, make sure to check over its work.

To better help train Greptile on our codebase, if the comment is useful and valid **Like** the comment, if its not helpful or invalid **Dislike**

To have Greptile Re-Review the changes, mention `greptileai`.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR introduces a layered cleanup mechanism for volume-browser helper containers: eager client-side teardown in `VolumeBrowser.svelte` (`onDestroy` + `pagehide` beacon), explicit "Stop browsing" button, pre-cleanup in `DeleteVolume`/`PruneVolumesWithOptions`, and a background reaper job (`PruningVolumeHelperJob`) as the final backstop.

- The `helperByVolume` map is promoted from `map[string]string` to `map[string]*volumeHelper`, adding a `lastUsedAt` timestamp. `StopHelper`, `ReapIdleHelpers`, and `touchHelperInternal` are carefully split so all Docker calls happen outside the mutex, with map entries pre-deleted before container removal — concurrent browse requests get a clean inspect-fail path and re-create a fresh helper.
- A new `VolumeBrowserHelperIdleTimeout` setting (default 10 minutes, 0 disables) is wired through models, defaults, the TypeScript type, and the schema test; the reaper job runs on a fixed 5-minute schedule and reads this knob each invocation.
- Pure-Go helper functions (`collectStaleHelperIDsInternal`, `takeHelperIDInternal`, `touchHelperInternal`) are tested in isolation without Docker, covering fresh/stale/edge/nil entries and idempotent absent cases.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

The change is safe to merge. All Docker calls are outside the mutex, force removal is used throughout, and concurrent browse/reap races result in a clean inspect-fail path rather than undefined behavior.

The concurrency design is sound: map entries are pre-deleted before container removal so concurrent requests transparently re-create a fresh helper, and force removal via `volumeHelperRemoveOptionsInternal` ensures running containers are stopped even when the timeout fires. DeleteVolume and PruneVolumes correctly stop helpers first to avoid "volume in use" conflicts. New unexported functions all carry the required `Internal` suffix. Unit tests cover edge cases (exact-timeout boundary, nil entries, idempotent take) without requiring Docker. The new setting is consistently plumbed through models, defaults, TypeScript types, and schema tests.

No files require special attention.
</details>

<sub>Reviews (7): Last reviewed commit: ["rename"](https://github.com/getarcaneapp/arcane/commit/d660beb4d88e2b3a10a8b6c74be9a1d1d677518d) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=34724168)</sub>

### CODE REVIEW

File: frontend/src/lib/components/file-browser/VolumeBrowser.svelte [hunk 1]
```diff
@@ -1,13 +1,53 @@
...
+		// async request during unload, so fire a keepalive beacon instead. Cookie auth
+		// (credentials are same-origin) means sendBeacon carries the session.
+		const handlePageHide = () => {
+			if (!envId) return;
+			const base = apiClient.defaults.baseURL || '/api';
+			navigator.sendBeacon(`${base}/environments/${envId}/volumes/${volumeName}/browse/stop`);
+		};
+		window.addEventListener('pagehide', handlePageHide);
+		return () => window.removeEventListener('pagehide', handlePageHide);
+	});
```
[the remark this answers is not in the document]
Comment (Zgrill2) [reply frontend/src/lib/components/file-browser/VolumeBrowser.svelte]: Fixed in recent push

File: backend/api/handlers/volumes.go [hunk 2]
```diff
@@ -501,6 +506,18 @@ func RegisterVolumes(api huma.API, dockerService *services.DockerClientService,
...
+		OperationID: "stop-volume-browser",
+		Method:      http.MethodPost,
+		Path:        "/environments/{id}/volumes/{volumeName}/browse/stop",
+		Summary:     "Stop volume browser helper",
+		Tags:        []string{"Volume Browser"},
+		Security: []map[string][]string{
+			{"BearerAuth": {}},
+			{"ApiKeyAuth": {}},
+		},
+	}, h.StopBrowsing)
```
[the remark this answers is not in the document]
Comment (Zgrill2) [reply backend/api/handlers/volumes.go]: Checked this. The endpoint can't stop arbitrary containers. StopBrowsing takes a volume name, not a container ID; StopHelper only removes the helper tracked in our internal helperByVolume map (the read-only container our own service created for that volume) and is a no-op otherwise. It's intentionally not admin-gated because it's part of the browse lifecycle, not the write tier. The read endpoints (browse/content/download) are already available to any authenticated user, and browsing is what spawns the helper. Gating stop behind admin would 403 the onDestroy/pagehide teardown for exactly those non-admin users who are allowed to browse, leaking helpers until the reaper. Worst case for a caller is forcing re-creation of a shared read-only helper (auto-recreated on the next request). If we want to restrict who can read volume contents, the right place is the read endpoints which is a separate decision from this PR.

**Comment by Zgrill2** [comment]
Fixed issues #2 and #3. Issue #1 about admin is a non-issue based on my analysis. The endpoint only allows stopping the helper container, not any arbitrary container and so does not need to be explicitly admin protected

**Comment by Zgrill2** [comment]
greptileai please re-review

**Comment by cabaucom376** [comment]
@greptileai

**Comment by Zgrill2** [comment]
@cabaucom376  do I have to do anything to get the E2E Test workflows to run? I ran all just tests on my local, is the report expected to upload or its running in a workflow?

please let me know if any additional changes requested

**Comment by Zgrill2** [comment]
@kmendell please review

File: backend/api/handlers/volumes.go [hunk 3]
```diff
@@ -501,6 +506,18 @@ func RegisterVolumes(api huma.API, dockerService *services.DockerClientService,
 		Middlewares: humamw.RequireAdmin(api),
 	}, h.DeleteFile)
 
+	huma.Register(api, huma.Operation{
```
Comment (kmendell) [inline backend/api/handlers/volumes.go]: I dont think youd need to api endpoint for this, just remove the temp container once its done, i thought i remember adding logic to prune them after like 5 minutes in case some other container needed to use them

**Review by kmendell — CHANGES_REQUESTED** [review]

File: backend/api/handlers/volumes.go (hunk 3 above)
Comment (Zgrill2) [reply backend/api/handlers/volumes.go]: Hi I will remove the frontend pieces and ensure the reaper runs in background every 5 minutes on idle containers. There is a process to prune these already existing, it only runs on update or shutdown

**Review by kmendell — APPROVED** [review]
